### PR TITLE
Add request-level tests for LspConnection

### PR DESCRIPTION
**Pull request by tintoy**
Hey - I'm starting to add more tests for the LSP client.

### CODE REVIEW

**Comment by tintoy** [comment]
@david-driscoll - would you mind having a look at `ClientTests.cs` and letting me know what you think? I could reduce the amount of boilerplate at the expense of more "magic", but I think maybe the first test I've added strikes a decent balance between boilerplate and magic 😁

**Comment by tintoy** [comment]
BTW, I can't see the AppVeyor build so I'll just have to assume it's working :)

**Comment by tintoy** [comment]
Failing, I meant failing 🤣

**Comment by david-driscoll** [comment]
I'll add you to the appveyor team

**Comment by david-driscoll** [comment]
And here's the error fyi:

```
 System.AggregateException : One or more errors occurred. (Error processing request '2' (500): Error processing request: Assert.Equal() Failure
                 (pos 0)
      Expected: \Foo.txt
      Actual:   
                 (pos 0)) (One or more errors occurred. (Safe handle has been closed))
      ---- OmniSharp.Extensions.LanguageServerProtocol.Client.LspRequestException : Error processing request '2' (500): Error processing request: Assert.Equal() Failure
                 (pos 0)
      Expected: \Foo.txt
      Actual:   
                 (pos 0)
      ---- System.AggregateException : One or more errors occurred. (Safe handle has been closed)
      -------- System.ObjectDisposedException : Safe handle has been closed
      Stack Trace:
        
        ----- Inner Stack Trace #1 (OmniSharp.Extensions.LanguageServerProtocol.Client.LspRequestException) -----
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
        C:\projects\csharp-language-server-protocol\src\Client\Protocol\LspConnection.cs(458,0): at OmniSharp.Extensions.LanguageServerProtocol.Client.Protocol.LspConnection.<SendRequest>d__38`1.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
           at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
        C:\projects\csharp-language-server-protocol\src\Client\Clients\TextDocumentClient.cs(79,0): at OmniSharp.Extensions.LanguageServerProtocol.Client.Clients.TextDocumentClient.<PositionalRequest>d__6`1.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
        C:\projects\csharp-language-server-protocol\test\Client.Tests\ClientTests.cs(86,0): at OmniSharp.Extensions.LanguageServerProtocol.Client.Tests.ClientTests.<Hover_Success>d__14.MoveNext()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
           at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
        ----- Inner Stack Trace #2 (System.AggregateException) -----
           at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
           at System.Threading.CancellationTokenSource.NotifyCancellation(Boolean throwOnFirstException)
           at System.Threading.CancellationTokenSource.Cancel(Boolean throwOnFirstException)
           at System.Threading.CancellationTokenSource.Cancel()
        C:\projects\csharp-language-server-protocol\src\Client\Protocol\LspConnection.cs(282,0): at OmniSharp.Extensions.LanguageServerProtocol.Client.Protocol.LspConnection.Disconnect(Boolean flushOutgoing)
        C:\projects\csharp-language-server-protocol\src\Client\Protocol\LspConnection.cs(171,0): at OmniSharp.Extensions.LanguageServerProtocol.Client.Protocol.LspConnection.Dispose()
           at System.Reactive.Disposables.CompositeDisposable.Dispose()
        C:\projects\csharp-language-server-protocol\test\Client.Tests\TestBase.cs(88,0): at OmniSharp.Extensions.LanguageServerProtocol.Client.Tests.TestBase.Dispose(Boolean disposing)
        C:\projects\csharp-language-server-protocol\test\Client.Tests\TestBase.cs(72,0): at OmniSharp.Extensions.LanguageServerProtocol.Client.Tests.TestBase.Dispose()
        C:\Dev\xunit\xunit\src\xunit.execution\Extensions\ReflectionAbstractionExtensions.cs(77,0): at ReflectionAbstractionExtensions.DisposeTestClass(ITest test, Object testClass, IMessageBus messageBus, ExecutionTimer timer, CancellationTokenSource cancellationTokenSource)
        ----- Inner Stack Trace -----
           at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
           at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
           at Interop.Kernel32.CancelIoEx(SafeHandle handle, NativeOverlapped* lpOverlapped)
           at System.IO.Pipes.PipeCompletionSource`1.Cancel()
           at System.IO.Pipes.PipeCompletionSource`1.<>c.<RegisterForCancellation>b__14_0(Object thisRef)
           at System.Threading.CancellationCallbackInfo.ExecutionContextCallback(Object obj)
           at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
           at System.Threading.CancellationCallbackInfo.ExecuteCallback()
           at System.Threading.CancellationTokenSource.CancellationCallbackCoreWork(CancellationCallbackCoreWorkArguments args)
           at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException
```

**Comment by tintoy** [comment]
Oops! Guess that wasn't so cross-platform after all (was on a mac when I tested it, so I guess the test will need to use a different path depending on OS). Will sort that out shortly :)

**Comment by tintoy** [comment]
BTW, have you seen Demystifier? Makes those stack traces a million times easier to read...

**Comment by tintoy** [comment]
@david-driscoll Here's a [Demystifier](https://github.com/benaadams/Ben.Demystifier) example: https://github.com/DimensionDataResearch/daas-demo/blob/master/src/DaaSDemo.Provisioning.Host/Program.cs#L36

It's only a test-level dependency, but I thought it'd be worth discussing before trying to add it (it's been pretty high-value in my own projects so far).

**Comment by david-driscoll** [comment]
I'd seen some stuff on twitter about it. I'm fine if you want to bring it in and make things look better.

**Comment by tintoy** [comment]
Cool - I'll look into it once the travis build starts working again (just want to be sure I've knocked the xplat paths issue on the head first).

**Comment by tintoy** [comment]
Hmm - looks like `keyserver.ubuntu.com` is down, the travis build keeps failing during the setup stage :-/

Will re-test on my mac.

**Comment by tintoy** [comment]
Ok - if the builds pass then this should be ready for review :)

File: test/Client.Tests/ConnectionTests.cs [hunk 1]
```diff
@@ -85,5 +85,165 @@ public ConnectionTests(ITestOutputHelper testOutput)
...
             await Task.WhenAll(clientConnection.HasHasDisconnected, serverConnection.HasHasDisconnected);
         }
+
+        /// <summary>
+        ///     Verify that a client <see cref="LspConnection"/> can handle a request from a server <see cref="LspConnection"/>.
+        /// </summary>
+        [Fact(DisplayName = "Client connection can handle request from server")]
+        public async Task Server_HandleRequest_Success()
+        {
+            LspConnection clientConnection = await CreateClientConnection();
```
Comment (david-driscoll) [inline test/Client.Tests/ConnectionTests.cs]: Minor nitpick, if we make `CreateClientConnection` return a disposable we can wrap all these calls in a using and remove some boilerplate with disconnect.

However that can be changed later.

**Pull request merged**